server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,485 @@
+                        queueRefs.decRef();
+                    }
+                } else {
+                    // request aborted, nothing more to send (queue is being cleared by queueRefs#closeInternal)
+                    isPartComplete = true;
+                    isLastPart = true;
+                    return new ReleasableBytesReference(BytesArray.EMPTY, () -> {});
+                }
+            } catch (Exception e) {
+                logger.error("failure encoding chunk", e);
+                throw e;
+            } finally {
+                if (targetStream != null) {
+                    assert false : "failure encoding chunk";
+                    IOUtils.closeWhileHandlingException(targetStream, Releasables.wrap(releasables));
+                    targetStream = null;
+                }
+            }
+        }
+
+        private void writeNextBytes(int sizeHint, Recycler<BytesRef> recycler, ArrayList<Releasable> releasables) throws IOException {
+            try {
+                if (bodyPart == null) {
+                    // When the last ref from listenersRefs is completed we enqueue a final null entry to trigger the transmission of the
+                    // zip file footer, which happens here:
+                    finishResponse(releasables);
+                    return;
+                }
+
+                if (zipEntry != null) {
+                    // This is the start of a new entry, so write the entry header:
+                    zipOutputStream.putNextEntry(zipEntry);
+                    zipEntry = null;
+                }
+
+                // Write the next chunk of the current entry to the zip stream
+                if (bodyPart.isPartComplete() == false) {
+                    try (var innerChunk = bodyPart.encodeChunk(sizeHint, recycler)) {
+                        final var iterator = innerChunk.iterator();
+                        BytesRef bytesRef;
+                        while ((bytesRef = iterator.next()) != null) {
+                            zipOutputStream.write(bytesRef.bytes, bytesRef.offset, bytesRef.length);
+                        }
+                    }
+                }
+                if (bodyPart.isPartComplete()) {
+                    // Complete the current part: if the current entry is incomplete then set up a listener for its next part, otherwise
+                    // move on to the next available entry and start sending its content.
+                    finishCurrentPart(releasables);
+                }
+            } finally {
+                // Flush any buffered data (but not the compressor) to chunkStream so that its size is accurate.
+                zipOutputStream.flush();
+            }
+        }
+
+        private void finishCurrentPart(ArrayList<Releasable> releasables) throws IOException {
+            if (bodyPart.isLastPart()) {
+                zipOutputStream.closeEntry();
+                transferCurrentEntryReleasable(releasables);
```

Comment:
In practice I think it can't, it can only propagate exceptions thrown from the underlying stream, and in this case the underlying stream is entirely in-memory and won't throw anything either. Note that we assert that no exceptions are thrown while `targetStream != null`. But even if it did throw, that'd still be safe, we keep a hold of the resource in `currentEntryReleasable` and release it in `drainQueue` on the cleanup path.